### PR TITLE
Make Gradient Module Divide by Distance Instead of Degrees

### PR DESCRIPTION
**Pull request by MO-PeterJordan**
The existing gradient module does not divide through by the distance between grid points. Instead it divides through by the angular separation, resulting in units of original-units * deg^-1. This PR changes that to divide through by distance in meters, so that the resulting units are original-units * m-1. This is what we would naturally expect, and allows eg. relative vorticity to be calculated with the correct units of s^-1.

Potential issues:
When calculating the gradient of wind speed (m/s) Iris presents the resulting units as Hz, which might be confusing for a gradient. However, grad_cube.units == s^1 returns true, so iris' units logic for downstream applications will be fine.

Absolute precision requirement has been reduced in the tests from e-8 to e-4 as this threshold was being breached in cases with very large gradients (>6 orders of magnitude between adjacent grid cells) where interpolation where required to recover the original grid shape . This seems to be to do with the way iris does linear interpolation. However O(10^6) gradients between adjacent cells seem unlikely to occur for meteorological parameters. Furthermore, +/-0.0001 should be enough precision for gradients of most if not all parameters.

### CODE REVIEW

**Review by mspelman07 — CHANGES_REQUESTED** [review]
Thanks for this Peter. The first Improver PR is always the hardest to get right.

I've added comments to the main plugins changes you've made. I haven't reviewed the tests yet but I felt I'd given you enough comments to keep you busy for now. I also notice there isn't a CLI for calculating the gradient. I don't know if that is part of this PR or will be part of another PR? We'll need it when we need to call the functions you've written from the suite.

**Review by MO-PeterJordan — COMMENTED** [review]
Most requested changes made. Some responded to with comments.

**Comment by MO-PeterJordan** [comment]
I think the CLI should be a separate PR, there's plenty on this one already!
I don't think I'd actually _need_ one for relative vorticity, as that will call the gradient module, which itself calls the distance module; however, it might be standard practice that all new modules get CLIs?

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I've suggested quite a lot of tweaks. Feel free to disagree where you feel strongly about it.

I haven't had time to finish reviewing all the tests - I'll do that when I re-review this.

File: improver_tests/utilities/test_DistanceBetweenGridPoints.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        "longitude",
+        longitudes,
+        "latitude",
+        latitudes,
+        "degrees",
+    )
+
+
+def test_latlon_cube_nonuniform_spacing():
+    input_cube = make_latlon_test_cube(
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver_tests/utilities/test_DistanceBetweenGridPoints.py:155]: done

File: improver_tests/utilities/test_DistanceBetweenGridPoints.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+        (expected_x_distances, expected_y_distances),
+    ):
+        assert result.units == "meters"
+        np.testing.assert_allclose(
+            result.data, expected.data, rtol=2e-3, atol=0
+        )  # Allowing 0.2% error for spherical earth approximation.
+
+
+def test_latlon_cube():
+    input_cube = make_latlon_test_cube(
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver_tests/utilities/test_DistanceBetweenGridPoints.py:215]: done

File: improver_tests/utilities/test_DistanceBetweenGridPoints.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    for result, expected in zip(
+        (calculated_x_distances_cube, calculated_y_distances_cube),
+        (expected_x_distances, expected_y_distances),
+    ):
+        assert result.units == "meters"
+        np.testing.assert_allclose(result.data, expected.data, rtol=2e-5, atol=0)
+
+
+def test_transverse_mercator_cube():
+    input_cube = make_transverse_mercator_test_cube((3, 2))
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver_tests/utilities/test_DistanceBetweenGridPoints.py:278]: done

File: improver_tests/utilities/test_DistanceBetweenGridPoints.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+        (expected_x_distances, expected_y_distances),
+    ):
+        assert result.units == "meters"
+        np.testing.assert_allclose(
+            result.data, expected.data, rtol=2e-3, atol=0
+        )  # Allowing 0.2% error for spherical earth approximation.
+
+
+def test_latlon_cube_unequal_xy_dims():
+    input_cube = make_latlon_test_cube(
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver_tests/utilities/test_DistanceBetweenGridPoints.py:185]: done

File: improver/utilities/spatial.py [hunk 5]
```diff
@@ -183,10 +187,248 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
+
+        Returns:
+            A cube containing the horizontal distances between the grid points of the input
+            cube in meters.
+        """
+        lats, longs = cls._get_latlon_cube_points(cube)
+
+        lats_as_col = np.expand_dims(lats, axis=1)
+        lon_diffs = np.diff(longs)
+        x_distances_degrees = np.tile(lon_diffs, (lats_as_col.shape[0], 1))
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Have simplified this method.

File: improver/utilities/spatial.py [hunk 6]
```diff
@@ -183,10 +187,248 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
+        Returns:
+            A cube containing the vertical distances between the grid points of the input
+            cube in meters.
+        """
+        lats, longs = cls._get_latlon_cube_points(cube)
+
+        lat_diffs = np.diff(lats)
+        y_distances_degrees = np.array(
+            [lat_diffs for _ in range(len(longs))]
+        ).transpose()
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Done

File: improver/utilities/spatial.py [hunk 7]
```diff
@@ -183,10 +187,248 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
+                Cube for which the distances are to be calculated.
+            x_diff:
+                Cube whose latitude and longitude dim coords match those of the cube to be output
+                by this method.
+
+        Returns:
+            A cube containing the vertical distances between the grid points of the input
+            cube in meters.
+        """
+        y_distances = calculate_grid_spacing(cube, axis="y", units="meters")
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Changed to y_grid_spacing

File: improver/utilities/spatial.py [hunk 8]
```diff
@@ -183,10 +187,248 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
+    @classmethod
+    def _get_distance_cube_y_distances(cls, cube: Cube, y_diff: Cube) -> Cube:
+        """
+        Calculates the vertical distances between adjacent grid points of a cube which uses
+        Equal Area coordinates.
+
+        Args:
+            cube:
+                Cube for which the distances are to be calculated.
+            x_diff:
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: I don't love this either, but struggling to think of anything better. The diff cubes are only there to provide coordinates with the correct dimensions to apply to the distances cubes. Open to suggestions for how to communicate this in a variable name though.

File: improver/utilities/spatial.py [hunk 9]
```diff
@@ -318,60 +560,50 @@ def __init__(self, regrid: bool = False) -> None:
...
         adjacent grid squares.
 
         Args:
             diff:
                 Cube containing differences along the x or y axis
-            axis:
-                Short-hand reference for the x or y coordinate, as allowed by
-                iris.util.guess_coord_axis.
+            name:
+                Name of the cube to be created.
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py:607]: Code deleted and below suggestion implemented to replace

File: improver/utilities/spatial.py [hunk 10]
```diff
@@ -183,10 +187,248 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
+
+        x_distances_meters = (
+            cls.EARTH_RADIUS
+            * np.cos(np.deg2rad(lats_full))
+            * np.deg2rad(x_distances_degrees)
+        )
+        dims = [(x_diff.coord("latitude"), 0), (x_diff.coord("longitude"), 1)]
+        return cls.build_distances_cube(x_distances_meters, dims, "x")
+
+    @classmethod
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Classmethods now all removed due to refactor.

File: improver/utilities/spatial.py [hunk 11]
```diff
@@ -318,60 +560,50 @@ def __init__(self, regrid: bool = False) -> None:
...
         Args:
             gradient:
                 Gradient values used in the data array of the resulting cube.
-            diff:
-                Cube containing differences along the x or y axis
-            cube:
-                Cube with correct output dimensions
-            axis:
-                Short-hand reference for the x or y coordinate, as allowed by
-                iris.util.guess_coord_axis.
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py:586]: Done

**Review by MO-PeterJordan — COMMENTED** [review]
Ready for re-review. Have implemented or replied to all review comments.
Also refactored the distance calculations to separate out the logic specific to different types of cubes a bit better. Hopefully makes it easier to follow.

File: improver/utilities/spatial.py [hunk 12]
```diff
@@ -255,141 +263,102 @@ def _get_latlon_cube_points(cube: Cube) -> Tuple[ndarray, ndarray]:
...
 
-        Args:
-            cube:
-                Cube for which the distances are to be calculated.
-            x_diff:
-                Cube whose latitude and longitude dim coords match those of the cube to be output
-                by this method.
-
         Returns:
             A cube containing the horizontal distances between the grid points of the input
```
Comment (MoseleyS) [inline improver/utilities/spatial.py:284]:
```suggestion
            A cube containing the x-axis distances between the grid points of the input
```

File: improver/utilities/spatial.py [hunk 13]
```diff
@@ -255,141 +263,102 @@ def _get_latlon_cube_points(cube: Cube) -> Tuple[ndarray, ndarray]:
...
-        Args:
-            cube:
-                Cube for which the distances are to be calculated.
-            x_diff:
-                Cube whose latitude and longitude dim coords match those of the cube to be output
-                by this method.
-
         Returns:
             A cube containing the vertical distances between the grid points of the input
             cube in meters.
```
Comment (MoseleyS) [inline improver/utilities/spatial.py]:
```suggestion
            A cube containing the y-axis distances between the grid points of the input
            cube in metres.
```

File: improver/utilities/spatial.py [hunk 14]
```diff
@@ -422,7 +387,56 @@ def process(self, cube: Cube, diffs: Tuple[Cube, Cube] = None) -> Tuple[Cube, Cu
...
+            return False
+
+    def process(self) -> Tuple[Cube, Cube]:
+        """
+        Calculate the distances between grid points along the x and y axes
+        and return the result in separate cubes.
+
+        Returns:
+            - Cube of horizontal distances.
+            - Cube of vertical distances.
```
Comment (MoseleyS) [inline improver/utilities/spatial.py:437]:
```suggestion
            - Cube of x-axis distances.
            - Cube of y-axis distances.
```

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Looking good, but I've found a few more things to tweak.

File: improver/utilities/spatial.py [hunk 15]
```diff
@@ -187,66 +187,74 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
-    Calculates the distances between adjacent grid squares within a cube.
-    The distances are calculated along the x and y axes individually.
-    Returned distances are in meters.
-    The class can handle cubes with either Geographic (lat-long) or Equal Area projections.
-    For lat-lon cubes, the distances are calculated assuming a spherical earth.
-    This causes a < 0.15% error compared with the full haversine equation.
-    """
-
-    EARTH_RADIUS = 6371e3  # meters
+class BaseDistanceCalculator:
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Done

File: improver/utilities/spatial.py (hunk 13 above)
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Fixed

File: improver/utilities/spatial.py [hunk 16]
```diff
@@ -187,66 +187,74 @@ def number_of_grid_cells_to_distance(cube: Cube, grid_points: int) -> float:
...
         """
-        coord_system = cube.coord_system()
-        return type(coord_system)
+        self.cube = cube
+        self.x_diff, self.y_diff = diffs
 
     @staticmethod
-    def _cube_xy_dimensions_are_distances(cube: Cube) -> bool:
+    def build_distances_cube(distances: ndarray, dims: List[Coord], axis: str) -> Cube:
+        """Constructs an output cube with units of meters."""
```
[the remark this answers is not in the document]
Comment (MO-PeterJordan) [reply improver/utilities/spatial.py]: Done

File: improver_tests/utilities/test_DistanceBetweenGridPoints.py [hunk 17]
```diff
@@ -58,13 +58,13 @@ def make_equalarea_test_cube(shape, grid_spacing, units="metres"):
...
     cube = set_up_variable_cube(
-        data, spatial_grid="equalarea", grid_spacing=grid_spacing
+        data, spatial_grid="equalarea", x_grid_spacing=grid_spacing, y_grid_spacing=grid_spacing
     )
     cube.coord("projection_x_coordinate").convert_units(units)
     cube.coord("projection_y_coordinate").convert_units(units)
     return cube
 
-
+#TODO: do I still need this now that set_up_variable_cube supports different x and y grid spacing?
```
Comment (MoseleyS) [inline improver_tests/utilities/test_DistanceBetweenGridPoints.py:67]: Added TODO?

File: improver_tests/utilities/test_GradientBetweenAdjacentGridSquares.py [hunk 18]
```diff
@@ -55,18 +57,26 @@
...
+X_GRID_SPACING_AT_10_DEGREES_NORTH = 13015378  # Metres/10 degrees * 12 = meters/120 degrees
+X_GRID_SPACING_AT_20_DEGREES_NORTH = 12113276  # Metres/10 degrees * 12 = meters/120 degrees # TODO: the * 12 is probaly unneccesarily confusing.
+Y_GRID_SPACING = 1111949  # Metres/10 degrees
 
 
-def regrid_x(data: np.ndarray) -> np.ndarray:
-    """
-    Regrids a 3 x 2 array to a 3 x 3 array using linear interpolation/extrapolation.
-    """
+def regrid_x_with_wrap_around_interpolation_for_edge_points(data):
```
Comment (MoseleyS) [inline improver_tests/utilities/test_GradientBetweenAdjacentGridSquares.py:69]: These methods need doc-strings too

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I see that this PR now has some conflicts with master, which prevents the CI tests from running. You'll need to sort this out before the PR can be approved.

**Comment by MoseleyS** [comment]
This work has been completed in other PRs.

**Pull request closed without merging**